File: src/main/java/com/lft/backend/dtos/UserDto.java
package com.lft.backend.dtos;

import com.lft.backend.User;
import com.lft.backend.enums.Sex;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Past;
import jakarta.validation.constraints.Size;

import java.time.LocalDate;
public class UserDto {
    @NotBlank(message = "First name is required")
    @Size(min = 1, max = 50, message = "First name must be between 1 and 50 characters")
    private String firstName;
    @NotBlank(message = "Last name is required")
    @Size(min = 1, max = 50, message = "Last name must be between 1 and 50 characters")
    private String lastName;
    @NotNull(message = "Date of birth is required")
    @Past(message = "Date of birth must be in the past")
    private LocalDate dateOfBirth;
    @Size(max = 30, message = "Nickname cannot exceed 30 characters")
    private String nickName;
    @NotNull(message = "Sex is required")
    private Sex sex;

    // Constructors
    public UserDto() {
    }

    public UserDto(String firstName, String lastName, LocalDate dateOfBirth, String nickName, Sex sex) {
        this.firstName = firstName;
        this.lastName = lastName;
        this.dateOfBirth = dateOfBirth;
        this.nickName = nickName;
        this.sex = sex;
    }

    // Static factory method: Entity → DTO
    public static UserDto from(User user) {
        if (user == null) return null;

        return new UserDto(
                user.getFirstName(),
                user.getLastName(),
                user.getDateOfBirth(),
                user.getNickName(),
                user.getSex()
        );
    }

    // Convert DTO → Entity
    public User toEntity() {
        return new User(firstName, lastName, dateOfBirth, nickName, sex);
    }

    // Getters and Setters
    public String getFirstName() {
        return firstName;
    }

    public void setFirstName(String firstName) {
        this.firstName = firstName;
    }

    public String getLastName() {
        return lastName;
    }

    public void setLastName(String lastName) {
        this.lastName = lastName;
    }

    public LocalDate getDateOfBirth() {
        return dateOfBirth;
    }

    public void setDateOfBirth(LocalDate dateOfBirth) {
        this.dateOfBirth = dateOfBirth;
    }

    public String getNickName() {
        return nickName;
    }

    public void setNickName(String nickName) {
        this.nickName = nickName;
    }

    public Sex getSex() {
        return sex;
    }

    public void setSex(Sex sex) {
        this.sex = sex;
    }
}

File: src/user/dtos/user.dto.ts
import {
  IsNotEmpty,
  IsString,
  MinLength,
  MaxLength,
  IsDateString,
  IsOptional,
  IsEnum,
} from 'class-validator';
import { Sex } from '../enums/sex.enum';
import { User } from '../entities/user.entity';

export class UserDto {
  @IsNotEmpty({ message: 'First name is required' })
  @IsString()
  @MinLength(1, { message: 'First name must be between 1 and 50 characters' })
  @MaxLength(50, { message: 'First name must be between 1 and 50 characters' })
  firstName: string;

  @IsNotEmpty({ message: 'Last name is required' })
  @IsString()
  @MinLength(1, { message: 'Last name must be between 1 and 50 characters' })
  @MaxLength(50, { message: 'Last name must be between 1 and 50 characters' })
  lastName: string;

  @IsNotEmpty({ message: 'Date of birth is required' })
  @IsDateString({}, { message: 'Date of birth must be a valid date' })
  dateOfBirth: string;

  @IsOptional()
  @IsString()
  @MaxLength(30, { message: 'Nickname cannot exceed 30 characters' })
  nickName?: string;

  @IsNotEmpty({ message: 'Sex is required' })
  @IsEnum(Sex, { message: 'Sex must be either MALE or FEMALE' })
  sex: Sex;

  constructor(
    firstName: string,
    lastName: string,
    dateOfBirth: string,
    nickName: string | null | undefined,
    sex: Sex,
  ) {
    this.firstName = firstName;
    this.lastName = lastName;
    this.dateOfBirth = dateOfBirth;
    this.nickName = nickName ?? undefined;
    this.sex = sex;
  }

  // Static factory method: Entity → DTO
  static from(user: User): UserDto | null {
    if (!user) return null;

    const dateOfBirth =
      user.dateOfBirth instanceof Date
        ? user.dateOfBirth.toISOString().split('T')[0]
        : user.dateOfBirth;

    return new UserDto(
      user.firstName,
      user.lastName,
      dateOfBirth,
      user.nickName,
      user.sex,
    );
  }

  // Convert DTO → Entity
  toEntity(): User {
    const user = new User();
    user.firstName = this.firstName;
    user.lastName = this.lastName;
    user.dateOfBirth = new Date(this.dateOfBirth);
    user.nickName = this.nickName ?? null;
    user.sex = this.sex;
    return user;
  }
}
